File: CodingInterviews/python/平衡二叉树.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def IsBalanced_Solution(self, pRoot):
        # write code here
        if pRoot is None:
            return True
        if not self.IsBalanced_Solution(pRoot.left) or not self.IsBalanced_Solution(pRoot.right):
            return False
        return abs(self.GetDepth(pRoot.left) - self.GetDepth(pRoot.right)) <= 1
    
    def GetDepth(self, pRoot):
        if pRoot is None:
            return 0 
        return max(self.GetDepth(pRoot.left), self.GetDepth(pRoot.right))+1

File: CodingInterviews/cpp/平衡二叉树.cpp
class Solution {
public:
    bool IsBalanced_Solution(TreeNode* pRoot) 
    {
        if(pRoot == NULL)
            return true;
        if(!(IsBalanced_Solution(pRoot->left) || !IsBalanced_Solution(pRoot->right)))
            return false;
        return abs(GetDepth(pRoot->left)-GetDepth(pRoot->right)) <= 1;
    }
    int GetDepth(TreeNode* pRoot)
    {
        if(pRoot == NULL)
            return 0;
        auto left = GetDepth(pRoot->left);
        auto right = GetDepth(pRoot->right);
        return left>right ? left+1:right+1;
    }
};



/*
 *solutions 2
 *  
*/
class Solution {
public:
    bool IsBalanced_Solution(TreeNode* pRoot) 
    {
        int h = 0;
        return IsBalanced(pRoot, h);
    }
    bool IsBalanced(TreeNode* pRoot, int &height) 
    {
        if(!pRoot) return true;
        int left=0,right=0;
        if(!IsBalanced(pRoot->left,left)) return false;
        if(!IsBalanced(pRoot->right,right)) return false;
        if(abs(left-right) > 1) return false;
        height = max(left,right) + 1;
        return true;
    }
};